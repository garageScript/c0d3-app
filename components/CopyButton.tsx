import React, { useState } from 'react'
import { Button } from './theme/Button'
import { AlertIcon, CheckIcon, CopyIcon } from '@primer/octicons-react'
import * as Sentry from '@sentry/react'

enum State {
  NotCopied,
  Copied,
  Error
}

enum ColorType {
  Primary = 'primary',
  Info = 'info',
  Success = 'success'
}

const CopyButton: React.FC<{ value: any }> = ({ value }: { value: any }) => {
  const [copyState, setCopyState] = useState<State>(State.NotCopied)
  const [type, setType] = useState<ColorType>(ColorType.Primary)

  const showSuccess = () => {
    setTimeout(() => {
      setCopyState(State.NotCopied)
      setType(ColorType.Primary)
    }, 2000)

    return <CheckIcon />
  }

  const handleOnClick = async () => {
    try {
      setType(ColorType.Primary)
      await navigator.clipboard.writeText(value)
      setCopyState(State.Copied)
      setType(ColorType.Success)
    } catch (e) {
      setCopyState(State.Error)
      setType(ColorType.Info)
      Sentry.captureException(e)
    }
  }

  return (
    <Button outline type={type} color={type} onClick={handleOnClick}>
      {copyState === State.Copied ? (
        showSuccess()
      ) : copyState === State.Error ? (
        <AlertIcon />
      ) : (
        <CopyIcon />
      )}
    </Button>
  )
}

export default CopyButton
