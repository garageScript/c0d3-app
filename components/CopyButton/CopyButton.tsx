import React, { useState } from 'react'
import { Button } from '../theme/Button'
import { AlertIcon, CheckIcon, CopyIcon } from '@primer/octicons-react'
import * as Sentry from '@sentry/react'
import { ColorTypes } from '../theme/colors'

enum State {
  NotCopied,
  Copied,
  Error
}

enum ColorType {
  Primary = 'primary',
  Info = 'info',
  Success = 'success',
  Mute = 'mute'
}

type Props = {
  value: string
  color?: 'success' | 'danger' | 'primary' | 'light' | 'info' | 'mute'
}

const CopyButton = ({ value, color }: Props) => {
  const bgColor = color || ColorType.Mute
  const [copyState, setCopyState] = useState<State>(State.NotCopied)
  const [type, setType] = useState(bgColor)

  const showSuccess = () => {
    setTimeout(() => {
      setCopyState(State.NotCopied)
      setType(bgColor)
    }, 2000)

    return <CheckIcon />
  }

  const handleOnClick = async () => {
    try {
      setType(bgColor)
      await navigator.clipboard.writeText(value)
      setCopyState(State.Copied)
      setType(ColorType.Success)
    } catch (e) {
      setCopyState(State.Error)
      setType(ColorType.Info)
      Sentry.captureException(e)
    }
  }

  return (
    <Button
      outline
      btnType={type}
      color={type as ColorTypes}
      onClick={handleOnClick}
    >
      {copyState === State.Copied ? (
        showSuccess()
      ) : copyState === State.Error ? (
        <AlertIcon />
      ) : (
        <CopyIcon />
      )}
    </Button>
  )
}

export default CopyButton
