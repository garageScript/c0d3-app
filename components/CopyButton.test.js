import { fireEvent, render, waitFor } from '@testing-library/react'
import React from 'react'
import { act } from 'react-dom/test-utils'
import CopyButton from './CopyButton'

Object.defineProperty(navigator, 'clipboard', {
  value: {
    writeText: () => {}
  },
  configurable: true
})
const testText = 'Test Text'

describe('Copy Button component', () => {
  test('Should render correctly', () => {
    const { container } = render(<CopyButton value={testText} />)

    expect(container).toMatchSnapshot()
  })

  test('Button type should be set to primary on render', () => {
    const { container } = render(<CopyButton value={testText} />)

    const button = container.querySelector('.btn')
    expect(button.classList.contains('btn-outline-bg-mute')).toBeTruthy()
  })

  test('Button type should be set to success on copy', async () => {
    const { container } = render(<CopyButton value={testText} />)

    const button = container.querySelector('.btn')

    fireEvent.click(button)
    await waitFor(() =>
      expect(button.classList.contains('btn-outline-bg-success')).toBeTruthy()
    )
  })

  test('Button type should be set to info on error', async () => {
    // Reset clipboard so writeText is undefined
    Object.defineProperty(navigator, 'clipboard', {
      value: null
    })

    const { container } = render(<CopyButton value={testText} />)

    const button = container.querySelector('.btn')

    fireEvent.click(button)
    await waitFor(() =>
      expect(button.classList.contains('btn-outline-bg-info')).toBeTruthy()
    )
  })

  test('Button type should be primary after 2 seconds on successful copy', async () => {
    jest.useFakeTimers()
    jest.spyOn(global, 'setTimeout')

    Object.defineProperty(navigator, 'clipboard', {
      value: {
        writeText: () => {}
      }
    })

    const { container } = render(<CopyButton value={testText} />)

    const button = container.querySelector('.btn')
    fireEvent.click(button)

    await waitFor(() =>
      expect(button.classList.contains('btn-outline-bg-success')).toBeTruthy()
    )

    act(() => jest.advanceTimersByTime(2000))

    await waitFor(() => {
      expect(button.classList.contains('btn-outline-bg-mute')).toBeTruthy()
      expect(setTimeout).toHaveBeenCalled()
    })
  })
})
