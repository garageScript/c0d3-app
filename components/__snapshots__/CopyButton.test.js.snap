// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copy Button component Should render correctly 1`] = `
<div>
  <button
    class="btn borderless btn-outline-mute onFocusBgFix btn-outline-bg-mute"
  >
    <svg
      aria-hidden="true"
      class="octicon octicon-copy"
      fill="currentColor"
      height="16"
      role="img"
      style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
      viewBox="0 0 16 16"
      width="16"
    >
      <path
        d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 010 1.5h-1.5a.25.25 0 00-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 00.25-.25v-1.5a.75.75 0 011.5 0v1.5A1.75 1.75 0 019.25 16h-7.5A1.75 1.75 0 010 14.25v-7.5z"
        fill-rule="evenodd"
      />
      <path
        d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0114.25 11h-7.5A1.75 1.75 0 015 9.25v-7.5zm1.75-.25a.25.25 0 00-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 00.25-.25v-7.5a.25.25 0 00-.25-.25h-7.5z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</div>
`;
